torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -0,0 +1,227 @@
+      nameMap_(std::move(nameMap)),
+      stop_(false),
+      pg_(std::move(pg)),
+      nextId_(0) {
+  TORCH_CHECK(nameMap_.size() > 1, "ProcessGroupAgent requires world_size to "
+      "be at least 2, but got ", nameMap_.size());
+  auto workerRankIter = nameMap_.find(workerName_);
+  TORCH_CHECK(workerRankIter != nameMap_.end(), "Failed to resolve worker "
+      "name ", workerName_, " to a ProcessGroup rank.");
+  TORCH_CHECK(pg_->getRank() == workerRankIter -> second,
+      "Resolved worker rank ", workerRankIter -> second,
+      " does not match ProcessGroup rank ", pg_->getRank());
+
+  names_.resize(nameMap_.size());
+  for (auto& entry : nameMap_) {
+    names_[entry.second] = entry.first;
+  }
+  sendThread_ = std::thread(&ProcessGroupAgent::sendLoop, this);
+  listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
+}
+
+void ProcessGroupAgent::join() {
+  // Every process i sends a SHUTDOWN message to process i + 1. This is
+  // necessary for now because:
+  // 1. There is no abort API for ProcessGroup::recvAnysource yet. We have to
+  //    feed it a message or kill the thread.
+  // 2. A GLOO process cannot send message to itself. (there is an ongoing
+  //    effort to fix this problem).
+  sync();
+  int dst = (pg_->getRank() + 1) % pg_->getSize();
+  enqueue(SendWork(dst, Message({}, {}, MessageType::SHUTDOWN)));
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  stop_ = true;
+  lock.unlock();
+
+  workProduceCV_.notify_all();
+  sendThread_.join();
+  listenerThread_.join();
+}
+
+void ProcessGroupAgent::sync() {
+  // Block until all processes wants to sync. This is necessary before acquiring
+  // the lock below, because other processes might not enter sync() until it
+  // gets some response from this RpcAgent.
+  pg_->barrier()->wait();
+  // Acquire the lock on the send queue to prevent additional messages to be put
+  // onto the send queue.
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+  // Wait until the send queue is depleted.
+  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  // Use another barrier in case different RpcAgent handles different amounts of
+  // workloads.
+  pg_->barrier()->wait();
+}
+
+std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+    const std::string& to, Message&& message) {
+
+  auto dstRankIter = nameMap_.find(to);
+  TORCH_CHECK(dstRankIter != nameMap_.end(), "Unknown destination worker ", to);
+
+  const int dstRank = dstRankIter -> second;
+  TORCH_CHECK(dstRank != pg_->getRank(), "ProcessGroupAgent does not support "
+    "making RPC calls to self.")
+
+  auto requestId = nextId();
+  auto future = std::make_shared<FutureMessage>();
+  if (message.isRequest()) {
+    {
+      std::lock_guard<std::mutex> lock{futureMutex_};
+      futures_[requestId] = future;
+    }
+    message.setId(requestId);
+  } else {
+    future->markCompleted();
```

Comment:
Is this mostly for the SHUTDOWN message? If so, we should have an assert here since its invalid to have a 'response' type message here.